File: lang/java/reef-bridge-proto-java/src/main/java/org/apache/reef/bridge/client/grpc/ClientService.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.reef.bridge.client.grpc;

import com.google.protobuf.ByteString;
import io.grpc.*;
import io.grpc.stub.StreamObserver;
import org.apache.reef.annotations.audience.ClientSide;
import org.apache.reef.bridge.client.grpc.parameters.ClientServerPort;
import org.apache.reef.bridge.proto.*;
import org.apache.reef.bridge.proto.Void;
import org.apache.reef.client.*;
import org.apache.reef.exception.NonSerializableException;
import org.apache.reef.runtime.common.utils.ExceptionCodec;
import org.apache.reef.tang.Configuration;
import org.apache.reef.tang.InjectionFuture;
import org.apache.reef.tang.annotations.Parameter;
import org.apache.reef.tang.annotations.Unit;
import org.apache.reef.util.Optional;
import org.apache.reef.wake.EventHandler;

import javax.inject.Inject;
import java.io.IOException;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Client service.
 */
@ClientSide
@Unit
final class ClientService extends REEFClientGrpc.REEFClientImplBase implements AutoCloseable {

  private static final Logger LOG = Logger.getLogger(ClientService.class.getName());

  private final Object waitLock = new Object();

  private final InjectionFuture<REEF> reef;

  private final ExceptionCodec exceptionCodec;

  private Server server = null;

  private BridgeClientGrpc.BridgeClientFutureStub clientStub = null;

  private LauncherStatus status = LauncherStatus.INIT;

  private RunningJob theJob = null;

  /**
   * This class should not be instantiated.
   */
  @Inject
  private ClientService(
      @Parameter(ClientServerPort.class) final int clientServerPort,
      final ExceptionCodec exceptionCodec,
      final InjectionFuture<REEF> reef) {
    this.reef = reef;
    this.exceptionCodec = exceptionCodec;
    if (clientServerPort > 0) {
      try {
        this.server =  ServerBuilder.forPort(0)
            .addService(this)
            .build()
            .start();
        final ManagedChannel channel = ManagedChannelBuilder
            .forAddress("localhost", clientServerPort)
            .usePlaintext()
            .build();
        this.clientStub = BridgeClientGrpc.newFutureStub(channel);
        this.clientStub.registerREEFClient(
            ClientProtocol.REEFClientRegistration.newBuilder()
                .setHostname("localhost")
                .setPort(this.server.getPort())
                .build());
      } catch (IOException e) {
        throw new RuntimeException("unable to start gRPC server", e);
      }
    }
  }

  @Override
  public void close() {
    this.reef.get().close();
    if (this.server != null) {
      this.server.shutdown();
      this.server = null;
    }
  }

  public LauncherStatus submit(final Configuration driverServiceConfiguration) {
    this.reef.get().submit(driverServiceConfiguration);
    synchronized (this.waitLock) {
      while (!this.status.isDone()) {
        try {
          LOG.log(Level.FINE, "Wait indefinitely");
          this.waitLock.wait();
        } catch (final InterruptedException ex) {
          LOG.log(Level.FINE, "Interrupted: {0}", ex);
        }
      }
    }
    return this.status;
  }

  @Override
  public void driverControlHandler(
      final ClientProtocol.DriverControlOp request,
      final StreamObserver<Void> responseObserver) {
    try {
      if (this.theJob != null && this.theJob.getId().equals(request.getJobId())) {
        try {
          switch (request.getOperation()) {
          case CLOSE:
            if (request.getMessage().isEmpty()) {
              this.theJob.close();
            } else {
              this.theJob.close(request.getMessage().toByteArray());
            }
            break;
          case MESSAGE:
            this.theJob.send(request.getMessage().toByteArray());
            break;
          default:
            throw new RuntimeException("Unknown operation " + request.getOperation());
          }
        } catch (Exception e) {
          responseObserver.onError(Status.INTERNAL
              .withDescription(e.getMessage())
              .asRuntimeException());
        } finally {
          responseObserver.onNext(Void.newBuilder().build());
        }
      } else {
        responseObserver.onError(Status.INTERNAL
            .withDescription("Unknown job " + request.getJobId())
            .asRuntimeException());
      }
    } finally {
      responseObserver.onCompleted();
    }
  }

  private void setStatusAndNotify(final LauncherStatus launcherStatus) {
    synchronized (this.waitLock) {
      this.status = launcherStatus;
      this.waitLock.notify();
    }
  }

  private ByteString getRootCause(final Throwable cause) {
    Throwable t = cause;
    LOG.log(Level.INFO, "Exception class {0}", t.getClass().getName());
    while (t.getCause() != null) {
      t = t.getCause();
    }
    // Do we bottom out at a non-serializable (e.g., C#) exception?
    if (t instanceof NonSerializableException) {
      LOG.log(Level.INFO, "non serializable exception");
      return ByteString.copyFrom(((NonSerializableException)t).getError());
    } else {
      LOG.log(Level.INFO, "serializable exception found");
      return ByteString.copyFrom(exceptionCodec.toBytes(t));
    }
  }

  /**
   * Job message handler.
   */
  final class JobMessageHandler implements EventHandler<JobMessage> {
    @Override
    public void onNext(final JobMessage value) {
      LOG.log(Level.INFO, "Job message from id {0}", value.getId());
      if (clientStub != null) {
        clientStub.jobMessageHandler(ClientProtocol.JobMessageEvent.newBuilder()
            .setJobId(value.getId())
            .setMessage(ByteString.copyFrom(value.get()))
            .build());
      }
    }
  }

  /**
   * Job driver notifies us that the job has been submitted to the Resource Manager.
   */
  final class SubmittedJobHandler implements EventHandler<SubmittedJob> {
    @Override
    public void onNext(final SubmittedJob job) {
      LOG.log(Level.INFO, "REEF job submitted: {0}.", job.getId());
      setStatusAndNotify(LauncherStatus.SUBMITTED);
      if (clientStub != null) {
        clientStub.jobSumittedHandler(ClientProtocol.JobSubmittedEvent.newBuilder()
            .setJobId(job.getId())
            .build());
      }
    }
  }

  /**
   * Job driver notifies us that the job is running.
   */
  final class RunningJobHandler implements EventHandler<RunningJob> {
    @Override
    public void onNext(final RunningJob job) {
      LOG.log(Level.INFO, "The Job {0} is running.", job.getId());
      theJob = job;
      setStatusAndNotify(LauncherStatus.RUNNING);
      if (clientStub != null) {
        clientStub.jobRunningHandler(ClientProtocol.JobRunningEvent.newBuilder()
            .setJobId(job.getId())
            .build());
      }
    }
  }

  /**
   * Job driver notifies us that the job had failed.
   */
  final class FailedJobHandler implements EventHandler<FailedJob> {
    @Override
    public void onNext(final FailedJob job) {
      final Optional<Throwable> ex = job.getReason();
      LOG.log(Level.SEVERE, "Received an error for job " + job.getId(), ex);
      theJob = null;
      setStatusAndNotify(LauncherStatus.failed(ex));
      if (clientStub != null) {
        final Optional<Throwable> error = exceptionCodec.fromBytes(job.getData().get());
        final ExceptionInfo exception;
        if (error.isPresent()) {
          LOG.log(Level.SEVERE, "Error for " + job.getId(), ex);
          exception = ExceptionInfo.newBuilder()
              .setName(job.toString())
              .setMessage(job.getMessage())
              .setData(getRootCause(error.get()))
              .build();
        } else {
          LOG.log(Level.SEVERE, "Unserialized error for job {0}: {1}",
              new Object[] {job.getId(), job.getMessage()});
          exception = ExceptionInfo.newBuilder()
              .setName(job.toString())
              .setMessage(job.getMessage())
              .build();
        }
        clientStub.jobFailedHandler(ClientProtocol.JobFailedEvent.newBuilder()
            .setJobId(job.getId())
            .setException(exception)
            .build());
      }
    }
  }

  /**
   * Job driver notifies us that the job had completed successfully.
   */
  final class CompletedJobHandler implements EventHandler<CompletedJob> {
    @Override
    public void onNext(final CompletedJob job) {
      LOG.log(Level.INFO, "The Job {0} is done.", job.getId());
      theJob = null;
      setStatusAndNotify(LauncherStatus.COMPLETED);
      if (clientStub != null) {
        clientStub.jobCompletedHandler(ClientProtocol.JobCompletedEvent.newBuilder()
            .setJobId(job.getId())
            .build());
      }
    }
  }

  /**
   * Handler an error in the job driver.
   */
  final class RuntimeErrorHandler implements EventHandler<FailedRuntime> {
    @Override
    public void onNext(final FailedRuntime error) {
      LOG.log(Level.SEVERE, "Received a resource manager error", error.getReason());
      theJob = null;
      setStatusAndNotify(LauncherStatus.failed(error.getReason()));
      if (clientStub != null) {
        clientStub.runtimeErrorHandler(ExceptionInfo.newBuilder()
            .setMessage(error.getMessage())
            .setName(error.getId())
            .setData(error.getReason().isPresent() ?
                getRootCause(error.getReason().get()) : ByteString.EMPTY)
            .build());
      }
    }
  }

  /**
   * Wake error handler.
   */
  final class WakeErrorHandler implements EventHandler<Throwable> {
    @Override
    public void onNext(final Throwable value) {
      LOG.log(Level.SEVERE, "Received a wake error", value);
      if (clientStub != null) {
        clientStub.wakeErrorHandler(ExceptionInfo.newBuilder()
            .setName(value.toString())
            .setMessage(value.getMessage())
            .setData(getRootCause(value))
            .build());
      }

    }
  }
}

File: lang/cs/Org.Apache.REEF.Bridge.Core.Grpc/Client/ClientService.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using Google.Protobuf;
using Grpc.Core;
using Org.Apache.REEF.Bridge.Core.Common.Client;
using Org.Apache.REEF.Bridge.Core.Common.Client.Config;
using Org.Apache.REEF.Bridge.Core.Common.Client.Events;
using Org.Apache.REEF.Bridge.Core.Proto;
using Org.Apache.REEF.Common.Client;
using Org.Apache.REEF.Tang.Annotations;
using Org.Apache.REEF.Tang.Exceptions;
using Org.Apache.REEF.Utilities;
using Org.Apache.REEF.Utilities.Logging;
using System;
using System.Threading;
using System.Threading.Tasks;
using Void = Org.Apache.REEF.Bridge.Core.Proto.Void;

namespace Org.Apache.REEF.Bridge.Core.Grpc.Client
{
    /// <summary>
    /// gRPC based client service.
    /// </summary>
    internal sealed class ClientService : BridgeClient.BridgeClientBase, IClientService
    {
        private static readonly Logger Log = Logger.GetLogger(typeof(ClientService));

        private static readonly Void Void = new Void();

        private readonly IObserver<ISubmittedJob> _submittedJobHandler;

        private readonly IObserver<IRunningJob> _runningJobHandler;

        private readonly IObserver<ICompletedJob> _completedJobHandler;

        private readonly IObserver<IFailedJob> _failedJobHandler;

        private readonly IObserver<IJobMessage> _jobMessageHandler;

        private readonly IObserver<IFailedRuntime> _failedRuntimeHandler;

        private readonly IObserver<IWakeError> _wakeErrorHandler;

        private REEFClient.REEFClientClient _clientStub = null;

        private string _jobId = "unknown";

        [Inject]
        private ClientService(
            [Parameter(Value = typeof(ClientParameters.SubmittedJobHandler))] IObserver<ISubmittedJob> submittedJobHandler,
            [Parameter(Value = typeof(ClientParameters.RunningJobHandler))] IObserver<IRunningJob> runningJobHandler,
            [Parameter(Value = typeof(ClientParameters.CompletedJobHandler))] IObserver<ICompletedJob> completedJobHandler,
            [Parameter(Value = typeof(ClientParameters.FailedJobHandler))] IObserver<IFailedJob> failedJobHandler,
            [Parameter(Value = typeof(ClientParameters.JobMessageHandler))] IObserver<IJobMessage> jobMessageHandler,
            [Parameter(Value = typeof(ClientParameters.FailedRuntimeHandler))] IObserver<IFailedRuntime> failedRuntimeHandler,
            [Parameter(Value = typeof(ClientParameters.WakeErrorHandler))] IObserver<IWakeError> wakeErrorHandler)
        {
            _submittedJobHandler = submittedJobHandler;
            _runningJobHandler = runningJobHandler;
            _completedJobHandler = completedJobHandler;
            _failedJobHandler = failedJobHandler;
            _jobMessageHandler = jobMessageHandler;
            _failedRuntimeHandler = failedRuntimeHandler;
            _wakeErrorHandler = wakeErrorHandler;
            LauncherStatus = LauncherStatus.InitStatus;
        }

        public bool IsDone => LauncherStatus.IsDone;

        public void Reset()
        {
            LauncherStatus = LauncherStatus.InitStatus;
        }

        public LauncherStatus LauncherStatus { get; private set; }

        public void Close(byte[] message = null)
        {
            try
            {
                _clientStub?.DriverControlHandler(new DriverControlOp()
                {
                    JobId = _jobId,
                    Message = message == null ? ByteString.Empty : ByteString.CopyFrom(message),
                    Operation = DriverControlOp.Types.Operation.Close
                });
            }
            catch (Exception e)
            {
                Log.Log(Level.Warning, "exception occurred when trying to close job", e);
            }
            LauncherStatus = LauncherStatus.ForceCloseStatus;
            _clientStub = null;
        }

        public void Send(byte[] message)
        {
            if (_clientStub != null)
            {
                _clientStub.DriverControlHandler(new DriverControlOp()
                {
                    JobId = _jobId,
                    Message = ByteString.CopyFrom(message),
                    Operation = DriverControlOp.Types.Operation.Message
                });
            }
            else
            {
                throw new IllegalStateException("Client service is closed");
            }
        }

        public override Task<Void> RegisterREEFClient(REEFClientRegistration request, ServerCallContext context)
        {
            Log.Log(Level.Info, "REEF Client registered on port {0}", request.Port);
            Channel driverServiceChannel = new Channel("127.0.0.1", (int)request.Port, ChannelCredentials.Insecure);
            _clientStub = new REEFClient.REEFClientClient(driverServiceChannel);
            return Task.FromResult(Void);
        }

        public override Task<Void> JobMessageHandler(JobMessageEvent request, ServerCallContext context)
        {
            Log.Log(Level.Info, "Job message from job id {0}", request.JobId);
            _jobMessageHandler.OnNext(new JobMessage(request.JobId, request.Message.ToByteArray()));
            return Task.FromResult(Void);
        }

        public override Task<Void> JobSumittedHandler(JobSubmittedEvent request, ServerCallContext context)
        {
            Log.Log(Level.Info, "Job id {0} submitted", request.JobId);
            UpdateStatusAndNotify(LauncherStatus.SubmittedStatus);
            _submittedJobHandler.OnNext(new SubmittedJob(request.JobId));
            _jobId = request.JobId;
            return Task.FromResult(Void);
        }

        public override Task<Void> JobRunningHandler(JobRunningEvent request, ServerCallContext context)
        {
            Log.Log(Level.Info, "Job id {0} running", request.JobId);
            UpdateStatusAndNotify(LauncherStatus.RunningStatus);
            _runningJobHandler.OnNext(new RunningJob(this, request.JobId));
            return Task.FromResult(Void);
        }

        public override Task<Void> JobCompletedHandler(JobCompletedEvent request, ServerCallContext context)
        {
            if (IsDone) return Task.FromResult(Void);
            Log.Log(Level.Info, "Job id {0} completed", request.JobId);
            UpdateStatusAndNotify(LauncherStatus.CompletedStatus);
            _completedJobHandler.OnNext(new CompletedJob(request.JobId));
            return Task.FromResult(Void);
        }

        public override Task<Void> JobFailedHandler(JobFailedEvent request, ServerCallContext context)
        {
            if (IsDone) return Task.FromResult(Void);
            Log.Log(Level.Info, "Job id {0} failed on {1}", request.JobId, request.Exception.Name);
            var jobFailedEvent = new FailedJob(request.JobId,
                request.Exception.Message,
                request.Exception.Data.ToByteArray());
            UpdateStatusAndNotify(LauncherStatus.Failed(jobFailedEvent.AsError()));
            _failedJobHandler.OnNext(jobFailedEvent);
            return Task.FromResult(Void);
        }

        public override Task<Void> RuntimeErrorHandler(ExceptionInfo request, ServerCallContext context)
        {
            if (!IsDone)
            {
                Log.Log(Level.Info, "Runtime error {0}", request.Message);
                UpdateStatusAndNotify(LauncherStatus.FailedStatus);
                _failedRuntimeHandler.OnNext(new FailedRuntime(_jobId, request.Message, request.Data.ToByteArray()));
            }

            return Task.FromResult(Void);
        }

        public override Task<Void> WakeErrorHandler(ExceptionInfo request, ServerCallContext context)
        {
            if (!IsDone)
            {
                Log.Log(Level.Info, "Wake error {0}", request.Message);
                UpdateStatusAndNotify(LauncherStatus.FailedStatus);
                _wakeErrorHandler.OnNext(new WakeError(_jobId,
                    request.Message,
                    Optional<byte[]>.Of(request.Data.ToByteArray())));
            }

            return Task.FromResult(Void);
        }

        private void UpdateStatusAndNotify(LauncherStatus status)
        {
            lock (this)
            {
                LauncherStatus = status;
                Monitor.PulseAll(this);
            }
        }
    }
}